### DISCUSSION TITLE
Recommended way of fetching async data inside re-used composable

### DISCUSSION THREAD

**Discussion by itzaks** [nuxt/nuxt#17608] [Questions] [answered]
Hi! 

Really enjoying Nuxt 3 so far. I have a question regarding how to fetch async data inside a custom `useXY`-composable

Currently I'm using `urql` to fetch graphql-data. I've enjoyed using `useAsyncData` in the components, but when using it in a composable, that I'm calling from other composables, I feel a bit unsure if I'm doing it the correct way. Will it be called multiple times if I'm calling upon `useDocuments` from multiple components and composable in the site? 

How do I best fetch this data once and re-use it in pages/components/composables?

```ts
export const useDocuments = () => {
  const { data, pending } = useAsyncData('all-documents', async () => {
    const urql = useClientHandle();
    const { data } = await urql.useQuery({
      query: DocumentsDataDocument,
    })
    return (data.value?.documents as Document[]) ?? []
  })

  return {
    data,
    pending
  }
}
```

I've also looked at some structure like this. But I feel I'm missing `useAsyncData`. Although, I could of course wrap it in this inside the page using it... I guess this is how to do it, or?

```ts
export const useDocuments = () => {
  const documents = useState<Document[]>('all-documents', [])
  const isLoading = ref(false)

  const fetch = async () => {
    isLoading.value = true
    const urql = useClientHandle();
    const { data } = await urql.useQuery({
      query: DocumentsDataDocument,
    })
    documents.value = (data.value?.documents as Document[]) ?? []
   }

  return {
    fetch,
    documents,
    isLoading
  }
}
```

That would look something like:

```ts
export const useDocuments = () => {
  const documents = useState<Document[]>('all-documents', [])

  const fetch = async () => {
    const urql = useClientHandle();
    const { data } = await urql.useQuery({
      query: DocumentsDataDocument,
    })
    documents.value = (data.value?.documents as Document[]) ?? []
   }

  return {
    fetch,
    documents,
  }
}
```

Using it like this:

```ts
const { fetch, documents } = useDocuments() 
const { pending } = useAsyncData('all-documents', async () => fetch())
```

However repeating this `'all-documents'` in multiple places feels a little un-DRY and I'd rather wrap it somewhere to be used once. No?

**Comment by andyjamesn** [comment]
Did you ever figure out the best approach for this?

**Reply by itzaks** [reply]
Nope, still unsure how to do this properly :–)

**Reply by avxkim** [reply]
i figured out, that we must use `useAsyncData` directly in a component without a wrapper, otherwise it doesn't work correctly.

**Comment by florian-lefebvre** [comment, marked as answer]
I believe your first snippet is okay, maybe making it async and tweaking it for types and names:
```ts
export const useDocuments = async () => {
const { data: documents, pending } = await useAsyncData('all-documents', async () => {
const urql = useClientHandle();
const { data } = await urql.useQuery({ query: DocumentsDataDocument });
return (data.value?.documents ?? []) as Document[];
})
return { documents, pending };
}
```
and then
```ts
const { documents, pending } = await useDocuments();
```

**Reply by avxkim** [reply]
If you're using a custom useAsyncData wrapper, do not await it in the composable, as that can cause unexpected behavior. Please follow [this recipe](https://nuxt.com/docs/guide/recipes/custom-usefetch#custom-usefetch) for more information on how to make a custom async data fetcher.

https://nuxt.com/docs/api/composables/use-async-data
